docs/book/v2/dispatching-middleware.md

```diff
@@ -0,0 +1,272 @@
+                     ],
+                 ],
+             ],
+       ],
+    ],
+];
+```
+
+## `PipeSpec` with PSR-15 Service Names
+
+Middleware can be passed as a `PipeSpec` object with container service class-string literals. This is the preferred
+method, if you need to dispatch multiple middleware instances. 
+
+Assuming you have two middleware, `Application\Middleware\SelectLanguageMiddleware` which determines the language of
+the visitor and `Application\Middleware\AlbumFromRouteMiddleware` which fetches an `Album` object by the ID in the route
+param. Both of them add this information to the `ServerRequestInterface` attributes which can be used by the
+`Application\Handler\AlbumDetailHandler` at the end of the pipe:
+
+```php
+use Application\Handler\AlbumDetailHandler;
+use Application\Middleware\AlbumFromRouteMiddlware;
+use Application\Middleware\SelectLanguageMiddleware;
+use Laminas\Mvc\Middleware\PipeSpec;
+use Laminas\Router\Http\Segment;
+
+return [
+    'router' => [
+        'routes' => [
+            'album-details' => [
+                'type' => Segment::class,
+                'options' => [
+                    'route' => '/album/:album_id',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new PipeSpec(
+                            SelectLanguageMiddleware::class,
+                            AlbumFromRouteMiddelware::class,
+                            AlbumDetailHandler::class
+                        ),
+                    ],
+                ],
+            ],
+        ],
+    ],
+];
+```
+
+##  Anonymous Interface Implementations
+
+You can also pass an anonymous class which implements `MiddlewareInterface` or `RequestHandlerInterface` directly.
+This is intended primarily for development because it can cause issues with configuration caching.
+
+```php
+use Laminas\Diactoros\Response\TextResponse;
+use Laminas\Mvc\Middleware\PipeSpec;
+use Laminas\Router\Http\Literal;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+return [
+    'router' => [
+        'routes' => [
+            'instance-middleware' => [
+                'type' => Literal::class,
+                'options' => [
+                    'route' => '/instance-middleware',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new class implements MiddlewareInterface {
+                            public function process(
+                                ServerRequestInterface $request,
+                                RequestHandlerInterface $handler
+                            ): ResponseInterface {
+                                // You have to return a response here because attempting to use $handler here will
```

Comment:
Double "here" seems redundant?
```suggestion
                                // You have to return a response directly because attempting to use $handler here will
```